firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,331 @@
+    velocity.  You must provide these with options using the glbaol
+    option ``Re`` for the Reynolds number and the prefixed option
+    ``pcd_velocity_space`` which should be the index into the full
+    space that gives velocity field.
+
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, \
+            assemble, inner, grad, split, Constant
+        prefix = pc.getOptionsPrefix() + "pcd_"
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Pressure space test and trial space differ")
+
+        Q = test.function_space()
+
+        p = TrialFunction(Q)
+        q = TestFunction(Q)
+
+        mass = p*q*dx
+
+        # Regularisation to avoid having to think about nullspaces.
+        stiffness = inner(grad(p), grad(q))*dx + Constant(1e-6)*p*q*dx
+
+        opts = PETSc.Options()
+        # we're inverting Mp and Kp, so default them to assembled.
+        # Fp only needs its action, so default it to mat-free.
+        # These can of course be overridden.
+        # only Fp is referred to in update, so that's the only
+        # one we stash.
+        Mp_mat_type = opts.getString(prefix+"Mp_mat_type", "")
+        Kp_mat_type = opts.getString(prefix+"Kp_mat_type", "")
+        self.Fp_mat_type = opts.getString(prefix+"Fp_mat_type", "matfree")
+
+        
+        # FIXME: allow these guys to be matrix-free
+        Mp = assemble(mass, form_compiler_parameters=context.fc_params,
+                      mat_type=Mp_mat_type)
+        Kp = assemble(stiffness, form_compiler_parameters=context.fc_params,
+                      mat_type=Kp_mat_type)
+
+        Mp.force_evaluation()
+        Kp.force_evaluation()
+
+        # FIXME: Should we transfer nullspaces over.  I think not.
+
+        Mksp = PETSc.KSP().create()
+        Mksp.setOptionsPrefix(prefix + "Mp_")
+        Mksp.setOperators(Mp.petscmat)
+        Mksp.setUp()
+        Mksp.setFromOptions()
+        self.Mksp = Mksp
+
+        Kksp = PETSc.KSP().create()
+        Kksp.setOptionsPrefix(prefix + "Kp_")
+        Kksp.setOperators(Kp.petscmat)
+        Kksp.setUp()
+        Kksp.setFromOptions()
+        self.Kksp = Kksp
+
+        state = context.context["state"]
+
+        Re = context.context.get("Re", 1.0)
+
+        velid = context.context["velocity_space"]
+
+        u0 = split(state)[velid]
+        fp = 1.0/Re * inner(grad(p), grad(q))*dx + inner(u0, grad(p))*q*dx
+
+        self.Re = Re
+        # FIXME, allow assembled matrix here
+        self.Fp = assemble(fp, form_compiler_parameters=context.fc_params,
+                           mat_type=self.Fp_mat_type)
+        self.Fp.force_evaluation()
+        Fpmat = self.Fp.petscmat
+        self.workspace = [Fpmat.createVecLeft() for i in (0, 1)]
```

Comment:
@pefarrell reminds me that the treatment of boundary conditions in PCD is somewhat subtle.  I think the best solution is in https://www.cs.umd.edu/~elman/papers/bc_approx_comm.pdf.  Given this, it suggests that we don't want PCD to be a builtin preconditioner (since handling all the correct inflow/outflow boundaries is rather problem dependent), e.g. dirichlet inflow conditions turn up as robin conditions on Fp.

Maybe we should instead incorporate this into the demos as an example "user-defined" preconditioner.

What do you think?